Add blog: 2023-02-05-github-slack (#240)
linkall-labs/docs: blog/2023-02-05-github-slack/index.md

vanus-controller-2               1/1     Running   0          69s
vanus-gateway-8677fc868f-rmjt9   1/1     Running   0          97s
vanus-store-0                    1/1     Running   0          96s
vanus-store-1                    1/1     Running   0          68s
vanus-store-2                    1/1     Running   0          68s
vanus-timer-5cd59c5bf-hmprp      1/1     Running   0          97s
vanus-timer-5cd59c5bf-pqkd5      1/1     Running   0          97s
vanus-trigger-7685d6cc69-8jgsl   1/1     Running   0          97s
```

4 Install vsctl (the command line tool)

```Plain
curl -O https://vsctl.s3.us-west-2.amazonaws.com/releases/v0.4.0/linux-amd64/vsctl
chmod ug+x vsctl
mv vsctl /usr/local/bin
```

5  Set the endpoint for vsctl to access vanus

```Plain
export VANUS_GATEWAY=192.168.49.2:30001
```

6 Create eventbus

```Plain
$ vsctl eventbus create  GitHub-slack
+----------------+-------------+
|     RESULT     |   EVENTBUS  |
+----------------+-------------+
| Create Success | github-slack|
+----------------+-------------+
```

## Part 2: Deploy the GitHub source connector

7 Set config file

 Create config.yml in any directory, the content is as follows

```Plain
{

>   "v_target": "http://192.168.49.2:30001/gateway/github-slack",

Author: > 

why 30002?

>   "v_target": "http://192.168.49.2:30001/gateway/github-slack",

Reviewer: This is the port, according to Vanus 0.6.0

  "v_port": "8082"
}
```

8 Deploy the GitHub source connector and run the following command in the same directory:

```Plain

> docker run --network=host -v $(pwd)/config.json:/vance/config/config.json  --rm vancehub/source-github > a.log &

Reviewer: ```suggestion
docker run -it --rm --network=host \
  -v ${PWD}:/vanus-connect/config \
  --name source-github public.ecr.aws/vanus/connector/source-github > a.log &
```

Reviewer: wrong command the connect was refractor

Author: 109:  -v ${PWD}:/vanus-connect/config  or -v ${PWD}:/vanus-connect/config.json

```

## Part 3: Creating a Slack app

10 Create a slack app

First, log in to slack, and click Create New APP,then select From Scrath, And fill in the App Name, select the corresponding Workspace

![create-slack-app](./img/create-slack-app.jpeg)

> 11 Select Incoming webhook，and click on

Reviewer: Creating an incoming webhook isnt needed for the slack sink. We should remove it.

![set-webhook-1](./img/set-webhook-1.jpeg)

![set-webhook-2](./img/set-webhook-2.jpeg)

![set-webhook-3](./img/set-webhook-3.jpeg)

13 Setting permissions

Select  **[OAuth & Permissions](https://api.slack.com/apps/A04L5D8QJ0Y/oauth?)** **,**  click Add an OAuth Scope in the Bot Token Scopes section of the Scope tab, and add chat:write and chat:write.public two types of permissions

![setting-permissions](./img/setting-permissions.jpeg)

Reinstall to Workspace

![reinstall](./img/reinstall.jpeg)

The slack app is created

## Part 4: Deploy the slack link connector

14 Create config.yml in any directory, the content is as follows

```Plain
default: "$app-name"
slack:
  - app_name: "$app-name"
    token: "$Bot User OAuth Token"
    default_channel: "$slack-channel"
```

A sample config.yml:

```Plain
default: "test"
slack:
  - app_name: "test"
    token: "xoxb-46xxxx7765-4685xxxx471152-DC5xxxxxyI1sj75rRUP"
    default_channel: "#vanus-channel"
```

15 Deploy the slack sink connector

```Plain

>  docker run  -p 31080:8080 --rm -v ${PWD}:/vance/config --name sink-slack public.ecr.aws/vanus/connector/sink-slack:latest > a.log & 

Reviewer: ```suggestion
 docker run -it --rm \
  -p 31080:8080 \
  -v ${PWD}:/vanus-connect/config \
  --name sink-slack public.ecr.aws/vanus/connector/sink-slack > a.log & 
```

- Through a filter, developers can filter out other events and only post the GitHub events they are interested in.
- Developers can process GitHub events through a transformer, extract key information from GitHub events, and arrange them according to their own needs.

​    Create a vanus subscription and set a filter or transformer in the subscription to achieve the above requirements. This article provides an example of event delivery for readers' reference.

   Get the event of Github star and post it to slack

   Create a subscription in vanus, and set up a transformer to extract and edit key information.

```Plain
vsctl subscription create  \
--eventbus github-slack  \
  --sink 'http://10.0.157.4:31080'  \
  --transformer '{
      "define": {
         "user":"$.data.sender.login",
         "time":"$.data.repository.updated_at",
         "homepage":"$.data.sender.html_url",
        "stargazers_count": "$.data.repository.stargazers_count",
        "repo": "$.data.repository.html_url"
      },

>       "template": "{\"subject\": \"${repo}\",\"message\":\"Hi Team, GitHub user < ${user} > just stared Vanus repository at  ${time} . Check out his GitHub home page here:${homepage}. We have ${stargazers_count} stars now !\"}"

Reviewer: ```suggestion
      "template": "{\"subject\": \"<repo>\",\"message\":\"Hi Team, GitHub user < <user> > just stared Vanus repository at  <time> . Check out his GitHub home page here: <homepage>. We have <stargazers_count> stars now !\"}"
```

Reviewer: This is according to the version update

 This article will help open-source enthusiasts deliver any Github events to Slack through Vanus in real-time. In this way, developers can know the status of open source projects in real-time without logging in to GitHub, so that developers can quickly respond to the following GitHub events:

This article will show how to do this in 5 minutes on Playground with Vanus and Vanus Connect， the results are shown below:

![GitHub-to-Slack-Result](./img/GitHub-to-Slack-Result.gif)

## Table of Contents

- What is GitHub
  - About GitHub
  - What GitHub events
  - Why need GitHub events
  - How to get github events
- What is Slack?
- How to Connect GitHub to Slack?
  - [Step 1: Deploying Vanus in the playground](# Step 1: Deploying Vanus in the playground)
  - [Step 2: Deploy the GitHub source connector](# Step 2: Deploy the GitHub source connector)
  - [Step 3: Creating a Slack app](# Step 3: Creating a Slack app)
  - [Step 4: Deploy the slack sink connector on Kubernetes](# Step 4: Deploy the slack sink connector on kubernetes)
- Test result
- [Summarize](# Summarize)

> 

Reviewer: ```suggestion
<!--truncate-->
```

- Comments event, PR comments, issue comments, commit comments.
- Stars event: A star is created or deleted from the repository.
- Version releases event:  Release created, edited, published, unpublished, or deleted.
- Wiki events: Wiki page updated.
- The team adds: Team added or modified on a repository.
- Discussions event:  created, edited, pinned, unpinned, locked, unlocked, transferred, answered, etc.
- Labels event: Label created, edited, or deleted.
- Milestone event: Milestone created, closed, opened, edited, or deleted.
- Code scanning alerts: Code Scanning alerts are created, fixed in a branch, or closed.

### Why need GitHub events

GitHub events provide an easy way to keep track of your GitHub repository without monitoring its status manually. They’re basically a notification system that offers a high level of customizability.

Through github events, you can learn a lot in real time, such as who starred the project, who submitted the PR, and whether a new version was released. At the same time, GitHub events can also trigger some operations, such as compiling code, automatic deployment, security checks, and so on.

## What is Slack

> <!--truncate-->

Reviewer: ```suggestion
```

Slack is an all-purpose communication platform and collaboration hub. It includes instant messaging, voice and video calls, and a suite of tools to help groups share information and work together.

> <!--truncate-->

Reviewer: ```suggestion

```

A Slack workspace is your team's home, similar to a dashboard.A Slack workspace is your team's home, similar to a dashboard.Slack Channels are shared group chat rooms for members of a workspace. Users can communicate with the entire team or certain team members in various channels.

> <!--truncate-->

Reviewer: ```suggestion

```

---
title: Build a notification system that pushes any Github event to Slack in 5 minutes
authors: Lucas
tags: ['GitHub', 'Message Queue', 'slack', 'notification']
---

## Build a notification system that pushes any GitHub event to Slack in 5 minutes

If you have an open-source project on GitHub, you definitely need to know who is attracted to your project in real-time. For example, whether someone has starred the project or submitted an Issue or a PR. How can we get the status of open-source projects in real-time? It is obviously not a good way to keep checking the GitHub page.

>  This article will help open-source enthusiasts deliver any Github events to Slack through Vanus in real-time. In this way, developers can know the status of open source projects in real-time without logging in to GitHub, so that developers can quickly respond to the following GitHub events:

Reviewer: ```suggestion
 This article will help open-source enthusiasts to deliver any Github events to Slack through Vanus in real-time. In this way, developers can know the status of open source projects in real-time without logging in to GitHub, so that developers can quickly respond to the following GitHub events:
```

This article will show how to do this in 5 minutes on Playground with Vanus and Vanus Connect， the results are shown below:

![GitHub-to-Slack-Result](./img/GitHub-to-Slack-Result.gif)

## Table of Contents

- What is GitHub
  - About GitHub

>   - What GitHub events

Reviewer: ```suggestion
  - What are GitHub events
```

- Wiki events: Wiki page updated.
- The team adds: Team added or modified on a repository.
- Discussions event:  created, edited, pinned, unpinned, locked, unlocked, transferred, answered, etc.
- Labels event: Label created, edited, or deleted.
- Milestone event: Milestone created, closed, opened, edited, or deleted.
- Code scanning alerts: Code Scanning alerts are created, fixed in a branch, or closed.

### Why need GitHub events

GitHub events provide an easy way to keep track of your GitHub repository without monitoring its status manually. They’re basically a notification system that offers a high level of customizability.

Through github events, you can learn a lot in real time, such as who starred the project, who submitted the PR, and whether a new version was released. At the same time, GitHub events can also trigger some operations, such as compiling code, automatic deployment, security checks, and so on.

## What is Slack
<!--truncate-->
Slack is an all-purpose communication platform and collaboration hub. It includes instant messaging, voice and video calls, and a suite of tools to help groups share information and work together.
<!--truncate-->

> A Slack workspace is your team's home, similar to a dashboard.A Slack workspace is your team's home, similar to a dashboard.Slack Channels are shared group chat rooms for members of a workspace. Users can communicate with the entire team or certain team members in various channels.

Reviewer: ```suggestion
 A Slack workspace is your team's home, similar to a dashboard .Slack Channels are shared group chat rooms for members of a workspace. Users can communicate with the entire team or certain team members in various channels.
```

vanus-controller-2               1/1     Running   0          69s
vanus-gateway-8677fc868f-rmjt9   1/1     Running   0          97s
vanus-store-0                    1/1     Running   0          96s
vanus-store-1                    1/1     Running   0          68s
vanus-store-2                    1/1     Running   0          68s
vanus-timer-5cd59c5bf-hmprp      1/1     Running   0          97s
vanus-timer-5cd59c5bf-pqkd5      1/1     Running   0          97s
vanus-trigger-7685d6cc69-8jgsl   1/1     Running   0          97s
```

4 Install vsctl (the command line tool)

```Plain
curl -O https://download.linkall.com/vsctl/latest/linux-amd64/vsctl
chmod ug+x vsctl
mv vsctl /usr/local/bin
```

5  Set the endpoint for vsctl to access vanus

```Plain
export VANUS_GATEWAY=192.168.49.2:30001
```

6 Create eventbus

```Plain
$ vsctl eventbus create  github-slack
+----------------+-------------+
|     RESULT     |   EVENTBUS  |
+----------------+-------------+
| Create Success | github-slack|
+----------------+-------------+
```

### Step 2: Deploy the GitHub source connector

1 Create webhook in GitHub repo

![create-webhook](./img/create-webhook.jpeg)

 **Payload URL** *

```

> http://ip10-1-53-4-cfie9skinko0oisrvrq0-8082.direct.play.linkall.com

Reviewer: This url changes for each users

Reviewer: Maybe we need to tell users how to get it from the playground

> http://ip10-1-53-4-cfie9skinko0oisrvrq0-8082.direct.play.linkall.com

Reviewer: ```suggestion
Get your payload URL in the GitHub to Twitter scenario under Payload URL.
Example: http://ip10-1-53-4-cfie9skinko0oisrvrq0-8082.direct.play.linkall.com
```

```

**Content type**

```
application/json
```

**Which events would you like to trigger this webhook?**

```
Send me everything.
```

2 Set config file

 Create config.yml in any directory, the content is as follows

```Plain
{
  "v_target": "http://192.168.49.2:30002/gateway/github-slack",
  "v_port": "8082"
}
```

3 Deploy the GitHub source connector and run the following command in the same directory:

```Plain
docker run -it --rm --network=host \
  -v ${PWD}:/vanus-connect/config \
  --name source-github public.ecr.aws/vanus/connector/source-github > a.log &
```

### Step 3: Creating a Slack app

1 Create a slack app

First, log in to slack, and click Create New APP,then select From Scrath, and fill in the App Name, select the corresponding Workspace

![create-slack-app](./img/create-slack-app.jpeg)

2 Setting permissions

Select  **[OAuth & Permissions](https://api.slack.com/apps/A04L5D8QJ0Y/oauth?)** **,**  click Add an OAuth Scope in the Bot Token Scopes section of the Scope tab, and add chat:write and chat:write.public two types of permissions

![setting-permissions](./img/setting-permissions.jpeg)

> Reinstall to Workspace

Reviewer: ```suggestion
Install to Workspace
```
